compiler/rustc_mir_transform/src/check_alignment.rs

```diff
@@ -0,0 +1,259 @@
+use crate::MirPass;
+use rustc_middle::mir::interpret::ConstValue;
+use rustc_middle::mir::interpret::Scalar;
+use rustc_middle::mir::*;
+use rustc_middle::ty::{TyCtxt, TypeAndMut};
+use rustc_span::DUMMY_SP;
+
+pub struct CheckAlignment;
+
+fn get_deref_place<'a>(place: &Place<'a>, rvalue: &Rvalue<'a>) -> Option<Place<'a>> {
+    if place.is_indirect() {
+        return Some(*place);
+    }
+    if let Rvalue::Use(Operand::Copy(place)) = rvalue {
+        if place.is_indirect() {
+            return Some(*place);
+        }
+    }
+    if let Rvalue::Use(Operand::Move(place)) = rvalue {
+        if place.is_indirect() {
+            return Some(*place);
+        }
+    }
+    if let Rvalue::Ref(_region, _kind, place) = rvalue {
+        if place.is_indirect() {
+            return Some(*place);
+        }
+    }
+    None
+}
+
+impl<'tcx> MirPass<'tcx> for CheckAlignment {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        // Skip this pass if debug-assertions are off
+        if !tcx.sess.opts.debug_assertions {
```

Comment:
oh 🤦 remove this so we can test it in perfbot?

Reply:
I think the perf run is informative, even with this condition in place. It shows no significant regression in the debug build times for a handful of crates, which means we aren't spending a lot of time emitting these checks.

However, I have now removed the condition in a separate commit. Fire away :)